site/docs/expressions/scalar_functions.md

```diff
@@ -43,10 +43,68 @@ For scalar functions (one value produced for each record), a function binding de
 
 
 
+## Argument Types & Output Type Derivation
+
+### Input Argument Types
+
+Input arguments can be declared in one of three ways: complete, wildcarded, parameterized.
+
+* Complete Type: Complete types are either simple types or compound types with all parameters fully defined. Examples include i32, fp32, VARCHAR(20), List&lt;fp32&gt;, etc.
+* Wildcard Types: Compound types can be wildcarded meaning that any valid value for the parameter can be accepted. Types are wildcarded by failing to express a particular value for the type. When serialized, this will typically mean using a null pointer or sentinel value. Examples would be  VARCHAR(*), List&lt;\*&gt;, DECIMAL(\*,10).
+* Parameterized types: Types can be parameterized such that the parameter can be used in a output derivation. Example parameterized types would be VARCHAR(T), List&lt;E&gt;, MAP&lt;K, fp32&gt;, etc. Parameters are simple names. Any parameters defined in an argument must be referenced at least once in an output expression. Parameterized argument types can only be used when a function is using expression based output type derivation.
+
+### Direct Return Types
+
+A direct return type is one that is fully known at function definition type. Example simple direct return types would be things such as i32, fp32. For compound types, a direct return type must be fully declared. Example fully defined types: VARCHAR(20), DECIMAL(25,5)
```

Comment:
> A direct return type is one that is fully known at function definition type.

Should the last `type` be `time`?

Also, I find this language confusing. A function definition with a parameterized type is fully known at function definition time. It won't necessarily have been monomorphized at definition time, is that what you're referring to here? Maybe I'm being a bit pedantic but my thinking is that we shouldn't invent new words for things :)

Reply:
Feel free to propose alternative language here.

Reply:
What is function definition time?  The point when the function is written down in the spec?  Or the point when the query plan is created?

If the former, perhaps simply:

A direct return type is one that does not depend on any parameterized input type.

If the latter:

A direct return type is one that is not know until the query has begun executing.

I assume from the context that we are talking about the former.

Reply:
Yes, the former. I'll update in next patch.